### PR TITLE
feat(start_planner): separate collision check and path planning

### PR DESCRIPTION
**Pull request by kyoichi-sugahara**
## Description

separate collision check feature from path planning function

## Tests performed

- [x] [PASS TIERIV INTERNAL SCENARIOS daily scenarios](https://evaluation.tier4.jp/evaluation/reports/ecbd4619-a188-5949-a5a2-1f87f9e45798?project_id=prd_jt)
- [x] [PASS TIERIV INTERNAL SCENARIOS common scenarios](https://evaluation.tier4.jp/evaluation/reports/0e6cbf96-8765-51d5-80c8-44cc5d7be776?project_id=prd_jt)

## Effects on system behavior

Not applicable.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_start_planner_module/src/start_planner_module.cpp [hunk 1]
```diff
@@ -616,43 +616,62 @@ PriorityOrder StartPlannerModule::determinePriorityOrder(
...
+  const auto & dynamic_objects = planner_data_->dynamic_object;
+  const auto pull_out_lanes = start_planner_utils::getPullOutLanes(
+    planner_data_, planner_data_->parameters.backward_path_length + parameters_->max_back_distance);
+  const auto & vehicle_footprint = createVehicleFootprint(vehicle_info_);
+  // extract stop objects in pull out lane for collision check
+  const auto [pull_out_lane_objects, others] =
+    utils::path_safety_checker::separateObjectsByLanelets(
+      *dynamic_objects, pull_out_lanes, utils::path_safety_checker::isPolygonOverlapLanelet);
+  const auto pull_out_lane_stop_objects = utils::path_safety_checker::filterObjectsByVelocity(
+    pull_out_lane_objects, parameters_->th_moving_object_velocity);
```
Comment (kosuke55) [inline planning/behavior_path_start_planner_module/src/start_planner_module.cpp:631]: doning filterObjectsByVelocity first may be faster

File: planning/behavior_path_start_planner_module/src/start_planner_module.cpp [hunk 2]
```diff
@@ -616,43 +616,62 @@ PriorityOrder StartPlannerModule::determinePriorityOrder(
...
+      combined_path.points.insert(
+        combined_path.points.end(), partial_path.points.begin(), partial_path.points.end());
+    }
+    // remove the point behind of shift end pose
+    const auto shift_end_pose_idx =
+      motion_utils::findNearestIndex(combined_path.points, path.end_pose);
+    combined_path.points.erase(
+      combined_path.points.begin() + *shift_end_pose_idx + 1, combined_path.points.end());
+    return combined_path;
+  };
```
Comment (kosuke55) [inline planning/behavior_path_start_planner_module/src/start_planner_module.cpp:658]: define in PullOutPath is also good

like PullOoverPath
https://github.com/autowarefoundation/autoware.universe/blob/feat/avoidance_pull_over/planning/behavior_path_goal_planner_module/include/behavior_path_goal_planner_module/pull_over_planner_base.hpp#L76-L87

File: planning/behavior_path_start_planner_module/src/start_planner_module.cpp (hunk 1 above)
Comment (kyoichi-sugahara) [reply planning/behavior_path_start_planner_module/src/start_planner_module.cpp:631]: agree! thanks!

File: planning/behavior_path_start_planner_module/src/start_planner_module.cpp [hunk 3]
```diff
@@ -616,43 +616,62 @@ PriorityOrder StartPlannerModule::determinePriorityOrder(
...
-  const auto next_pull_out_path = planner->plan(next_pull_out_start_pose, goal_pose);
+  // If start pose candidate
+  if (backward_is_unnecessary) {
+    updateStatusWithCurrentPath(*pull_out_path, start_pose_candidate, planner->getPlannerType());
+    return true;
+  }
 
-  // If no next path is found, return false
-  if (!next_pull_out_path) return false;
+  updateStatusWithNextPath(*pull_out_path, start_pose_candidate, planner->getPlannerType());
```
[the remark this answers is not in the document]
Comment (kyoichi-sugahara) [reply planning/behavior_path_start_planner_module/src/start_planner_module.cpp:674]: next_pull_out_path for avoiding replan of backward path after backward driving.
But replan after backward driving doesn't happen, so remove next_pull_out_path

File: planning/behavior_path_start_planner_module/src/start_planner_module.cpp (hunk 1 above)
Comment (kyoichi-sugahara) [reply planning/behavior_path_start_planner_module/src/start_planner_module.cpp]: changed in https://github.com/autowarefoundation/autoware.universe/pull/5952/commits/e5717adb0d8235d7393240ef3b3fbd53150c48bb

**Review by kosuke55 — APPROVED** [review]

**Pull request merged**